src/com/amazon/ionschema/internal/SchemaImpl.kt

```diff
@@ -129,35 +126,52 @@ internal class SchemaImpl private constructor(
                            header: IonStruct): Map<String, Import> {
 
         val importsMap = mutableMapOf<String, SchemaAndTypeImports>()
+        val importSet: MutableSet<String> = schemaSystem.getSchemaImportSet()
+
         (header.get("imports") as? IonList)
             ?.filterIsInstance<IonStruct>()
             ?.forEach {
-                val id = it["id"] as IonText
-                val importedSchema = schemaSystem.loadSchema(id.stringValue())
-                val schemaAndTypes = importsMap.getOrPut(id.stringValue()) {
-                    SchemaAndTypeImports(id.stringValue(), importedSchema)
-                }
+                val childImportId = it["id"] as IonText
+                val alias = it["as"] as? IonSymbol
+                // if importSet has an import with this id then do not load schema again to break the cycle.
+                if(!importSet.contains(childImportId.stringValue())) {
+                    var parentImportId = schemaId?: "";
+
+                    // if Schema is importing itself then throw error
+                    if(parentImportId.equals(childImportId.stringValue())) {
+                        throw InvalidSchemaException("Schema can not import itself.")
+                    }
```

Comment:
Interesting corner case here: given the intentionally open-ended language in the spec, I think it's possible for the empty string to be a valid import `id`. `parentImportId` can be set to `""` on line 138; if that happens, attempting to import a child with `id=""` would cause an error.

To understand this better we'd need to look at all the invocation sites of this class's constructor to see where/why `schemaId` is set to `null`.

This doesn't need to hold up the PR, I'm adding this comment so I can easily open an issue for this. :)

Reply:
The grammar does indeed have `ID` as either a `STRING` or `SYMBOL` both of which allow for the empty value. That is unfortunate in my opinion :)